### PR TITLE
merge utf8 to master

### PR DESCRIPTION
**Pull request by anuchak**
utf8 changes to be merged with master

### CODE REVIEW

File: dataset/tinysnb/schema.cypher [hunk 1]
```diff
@@ -1,5 +1,6 @@
 create node table person (ID INt64, fName StRING, gender INT64, isStudent BoOLEAN, isWorker BOOLEAN, age INT64, eyeSight DOUBLE, birthdate DATE, registerTime TIMESTAMP, lastJobDuration interval, workedHours INT64[], usedNames STRING[], courseScoresPerTerm INT64[][], PRIMARY KEY (ID));
 create node table organisation (ID INT64, name STRING, orgCode INT64, mark DOUBLE, score INT64, history STRING, licenseValidInterval INTERVAL, rating DOUBLE, PRIMARY KEY (ID));
+create node table movies (ID INT64, name STRING, PRIMARY KEY (ID));
```
Comment (acquamarin) [inline dataset/tinysnb/schema.cypher:3]: I think this dataset is only used to test utf-8 strings. To make the testing simpler, you can have one simply column name and use name as primary key.

File: third_party/utf8proc/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+cc_library(
+    name = "utf8proc",
+    srcs = glob([
+        "*.cpp",
+    ]),
+    hdrs = glob([
+        "include/*.h",
+    ]),
+    visibility = ["//visibility:public"],
+    deps = [],
```
Comment (acquamarin) [inline third_party/utf8proc/BUILD.bazel:12]: If this library doesn't have any deps, you don't need to write deps = [].

File: src/function/string/operations/include/base_lower_upper_operation.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            auto buffer = reinterpret_cast<char*>(result.overflowPtr);
+            convertCase(buffer, input.len, (char*)input.getData(), isUpper);
+            memcpy(result.prefix, buffer, ku_string_t::PREFIX_LENGTH);
+        }
+    }
+
+private:
+    static uint32_t getResultLen(char* inputStr, uint32_t inputLen, bool isUpper);
+    static void convertCase(char* result, uint32_t len, char* input, bool toUpper);
+};
```
Comment (acquamarin) [inline src/function/string/operations/include/base_lower_upper_operation.h:38]: Newline here.

File: src/function/string/operations/include/lpad_operation.h [hunk 4]
```diff
@@ -3,33 +3,32 @@
...
 using namespace std;
-using namespace kuzu::common;
+using namespace kuzu::function::operation;
 
 namespace kuzu {
 namespace function {
 namespace operation {
 
-struct Lpad : PadOperation {
+struct Lpad : BasePadOperation {
```
Comment (acquamarin) [inline src/function/string/operations/include/lpad_operation.h:17]: lpad will cause an assertion failure if the count is a negative number:
```
return lpad('5', -1, '>')
```

File: test/test_files/tinySNB/function/string.test [hunk 5]
```diff
@@ -318,6 +339,13 @@ Faroo
 <Greg
 Huber
 
+-NAME LpadUTF8String
+-QUERY MATCH (m:movies) RETURN lpad(m.name, 25, "âó")
+---- 3
+âóâóâóâSóló cón tu párejâ
+âóâThe 😂😃🧘🏻‍♂️🌍🌦️🍞🚗 movie
+âóâóâóâóâóâóâóâóâóâóâRoma
+
```
Comment (acquamarin) [inline test/test_files/tinySNB/function/string.test:348]: Can you add tests for lpad/rpad negative idx as well?

**Review by acquamarin — CHANGES_REQUESTED** [review]
1. There is a bug in lpad/rpad operation when the count is a negative number. Also add tests to cover these two cases.
2. Shell bug:
a. If we are typing utf-8 strings in the shell, the ctrl-left,ctrl-right is not working. 
b. If we are typing utf-8 strings in the shell, the number of characters that can fit in the screen seems incorrect:
If you are typing a long utf-8 string in the shell, the left most utf-8 characters will be hidden even if there are still spaces left in the screen.

File: third_party/utf8proc/BUILD.bazel (hunk 2 above)
Comment (anuchak) [reply third_party/utf8proc/BUILD.bazel]: removed

File: src/function/string/operations/include/base_lower_upper_operation.h (hunk 3 above)
Comment (anuchak) [reply src/function/string/operations/include/base_lower_upper_operation.h:38]: done

File: src/function/string/operations/include/lpad_operation.h (hunk 4 above)
Comment (anuchak) [reply src/function/string/operations/include/lpad_operation.h:17]: changed this to same as duckdb behaviour, they change -ve count to 0

File: test/test_files/tinySNB/function/string.test (hunk 5 above)
Comment (anuchak) [reply test/test_files/tinySNB/function/string.test:348]: done

File: dataset/tinysnb/schema.cypher (hunk 1 above)
Comment (anuchak) [reply dataset/tinysnb/schema.cypher:3]: Yes previously ID didn't support chars so had to add that.
I think 2 columns are fine, or else I'll have to go back and change all the tests for this.

Comment (acquamarin) [reply dataset/tinysnb/schema.cypher:3]: We always use the minimum code/dataset for testing, so it is better for you to just keep one column in movies table.

**Review by acquamarin — CHANGES_REQUESTED** [review]
I want to take another look after you fixed the bug in left & right

**Review by acquamarin — COMMENTED** [review]
```
kuzu> match (t:test) return t;
----------
| t.name |
----------
| alice  |
----------
| 人    |
----------

```
I think we need to change the result printer as well. If it is easy, you can do in your current PR. 
Otherwise, open an issue for this, and let aziz handle this.

File: src/function/string/operations/include/left_operation.h [hunk 6]
```diff
@@ -18,8 +20,9 @@ struct Left {
 public:
     static inline void operation(
         ku_string_t& left, int64_t& right, ku_string_t& result, ValueVector& resultValueVector) {
-        auto len = right >= 0 ? min(left.len, (uint32_t)right) :
-                                ((uint32_t)max(left.len + right, (int64_t)0));
+        int64_t leftLen;
+        Length::operation(left, leftLen);
+        int64_t len = (right > 0) ? min(leftLen, right) : max(leftLen + right, (int64_t)0);
```
[the remark this answers is not in the document]
Comment (anuchak) [reply src/function/string/operations/include/left_operation.h]: fixed the issue, it is printing empty val now.

File: src/function/string/operations/include/length_operation.h [hunk 7]
```diff
@@ -3,17 +3,38 @@
...
 
 struct Length {
-    static inline void operation(ku_string_t& input, int64_t& result) { result = input.len; }
+    static inline void operation(ku_string_t& input, int64_t& result) {
+        auto totalByteLength = input.len;
+        auto inputString = input.getAsString();
+        for (auto i = 0; i < totalByteLength; i++) {
+            if (inputString[i] & 0x80) {
+                int64_t length = 0;
+                // use grapheme iterator to identify bytes of utf8 char and increment once for each
```
[the remark this answers is not in the document]
Comment (anuchak) [reply src/function/string/operations/include/length_operation.h]: done

File: src/function/string/operations/include/right_operation.h [hunk 8]
```diff
@@ -18,9 +19,10 @@ struct Right {
 public:
     static inline void operation(
         ku_string_t& left, int64_t& right, ku_string_t& result, ValueVector& resultValueVector) {
-        auto len = right >= 0 ? min(left.len, (uint32_t)right) :
-                                ((uint32_t)max((int64_t)0, left.len + right));
-        SubStr::operation(left, left.len - len + 1, len, result, resultValueVector);
+        int64_t leftLen;
+        Length::operation(left, leftLen);
+        int64_t len = (right > 0) ? min(leftLen, right) : max(leftLen + right, (int64_t)0);
+        SubStr::operation(left, leftLen - len + 1, len, result, resultValueVector);
```
[the remark this answers is not in the document]
Comment (anuchak) [reply src/function/string/operations/include/right_operation.h:25]: fixed it

File: src/function/string/operations/include/upper_operation.h [hunk 9]
```diff
@@ -18,15 +16,7 @@ struct Upper {
...
-        BaseStrOperation::operation(input, result, resultValueVector, upperStr);
-    }
-
-private:
-    static uint32_t upperStr(char* str, uint32_t len) {
-        for (auto i = 0u; i < len; i++) {
-            str[i] = toupper(str[i]);
-        }
-        return len;
+        BaseLowerUpperOperation::operation(input, result, resultValueVector, /* isUpper */ true);
```
[the remark this answers is not in the document]
Comment (anuchak) [reply src/function/string/operations/include/upper_operation.h]: changed

**Comment by anuchak** [comment]
I've fixed the left & right operation bugs.

For the result printer, I tried to fix it but can't find the solution for an issue, even if we count `A` and `大` as 1 character, when printing them, the later takes up more space than `A`. We can use the utf8proc library to count chars properly, but it provides no facility find how much space a utf8 character takes up.

I've raised an issue here #1047

**Comment by anuchak** [comment]
I ran some queries with duckdb for long utf8 characters and it seems they don't print the characters.
They just cut it short with ellipsis `...`

File: dataset/tinysnb/schema.cypher (hunk 1 above)
Comment (anuchak) [reply dataset/tinysnb/schema.cypher]: I've dropped the extra column, now name is the only column and primary key.

**Review by acquamarin — CHANGES_REQUESTED** [review]
```
kuzu> return array_extract('都是大叔大的撒的', 4);
------------------------------------------------
| array_extract('都是大叔大的撒的', 4) |
------------------------------------------------
| ?                                            |
------------------------------------------------
(1 tuple)
Time: 14.64ms (compiling), 0.70ms (executing)
kuzu> 

```
Have you checked the `array_extract` function? Please be patient and check the correctness every string functions carefully. All functions that supported by our system are under: `src/function/string/operations/include`.
BTW: we also need to update the `find` operation to make it compatible with UTF8.

**Comment by anuchak** [comment]
Both list_extract and array_extract didn't have UTF8 char support, that's why they were not working.
I've fixed both of them:

------------------------------------------------
| array_extract('都是大叔大的撒的', 4) |
------------------------------------------------
| 叔                                          |
------------------------------------------------
(1 tuple)
Time: 19.24ms (compiling), 1.39ms (executing)

-----------------------------------------------
| list_extract('都是大叔大的撒的', 4) |
-----------------------------------------------
| 叔                                         |
-----------------------------------------------
(1 tuple)
Time: 8.62ms (compiling), 0.60ms (executing)


For the `find` function, not sure which function are you referring to ? There is no such function in the catalog currently.
Also I checked DuckDB's find function code (the needle haystack algorithm) and they have no UTF8 specific code there.

File: tools/shell/linenoise.cpp [hunk 10]
```diff
@@ -953,11 +1037,15 @@ static int linenoiseEdit(
             break;
         case CTRL_T: /* ctrl-t, swaps current character with previous. */
             if (l.pos > 0 && l.pos < l.len) {
-                int aux = buf[l.pos - 1];
-                buf[l.pos - 1] = buf[l.pos];
-                buf[l.pos] = aux;
-                if (l.pos != l.len - 1)
-                    l.pos++;
+                char tempBuffer[128];
```
Comment (acquamarin) [inline tools/shell/linenoise.cpp:1040]: we always use `tmp` instead of `temp` for variable names

**Review by acquamarin — APPROVED** [review]

File: tools/shell/linenoise.cpp (hunk 10 above)
Comment (anuchak) [reply tools/shell/linenoise.cpp]: renamed it
